persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateQueryHandler.java

```diff
@@ -64,15 +87,188 @@ public ResultMessage process(
+                castStatement.keyspace(), castStatement.columnFamily()));
+      }
+
+      logger.debug(
+          "authorized statement of type {} on {}.{}",
+          castStatement.getClass().toString(),
+          castStatement.keyspace(),
+          castStatement.columnFamily());
+    } else if (statement instanceof SchemaAlteringStatement) {
+      SchemaAlteringStatement castStatement = (SchemaAlteringStatement) statement;
+      Scope scope = null;
+      String keyspaceName = null;
+      String tableName = null;
+
+      if (statement instanceof CreateTableStatement) {
+        scope = Scope.CREATE;
+        keyspaceName = castStatement.keyspace();
+        tableName = castStatement.columnFamily();
+      } else if (statement instanceof DropTableStatement) {
+        scope = Scope.DELETE;
+        keyspaceName = castStatement.keyspace();
+        tableName = castStatement.columnFamily();
+      } else if (statement instanceof AlterTableStatement) {
+        scope = Scope.ALTER;
+        keyspaceName = castStatement.keyspace();
+        tableName = castStatement.columnFamily();
+      } else if (statement instanceof CreateKeyspaceStatement) {
+        scope = Scope.CREATE;
+        keyspaceName = castStatement.keyspace();
+        tableName = null;
+      } else if (statement instanceof DropKeyspaceStatement) {
+        scope = Scope.DELETE;
+        keyspaceName = castStatement.keyspace();
+        tableName = null;
+      } else if (statement instanceof AlterKeyspaceStatement) {
+        scope = Scope.ALTER;
+        keyspaceName = castStatement.keyspace();
+        tableName = null;
+      }
+
+      logger.debug(
+          "preparing to authorize statement of type {} on {}.{}",
+          castStatement.getClass().toString(),
+          keyspaceName,
+          tableName);
+
+      try {
+        authorization.authorizeSchemaWrite(authToken, keyspaceName, tableName, scope);
+      } catch (io.stargate.auth.UnauthorizedException e) {
+        throw new UnauthorizedException(
+            String.format(
+                "Missing correct permission on %s.%s",
+                keyspaceName, (tableName == null ? "" : tableName)));
+      }
+
+      logger.debug(
+          "authorized statement of type {} on {}.{}",
+          castStatement.getClass().toString(),
+          keyspaceName,
+          tableName);
+    }
+  }
+
+  private AuthorizationService getAuthorizationServiceFromContext() {
+    BundleContext bundleContext = FrameworkUtil.getBundle(this.getClass()).getBundleContext();
+
+    ServiceReference<?>[] refs;
+    try {
+      refs = bundleContext.getServiceReferences(AuthorizationService.class.getName(), null);
+    } catch (InvalidSyntaxException e) {
+      logger.error("Failed to get service reference for AuthorizationService", e);
+      return null;
+    }
+
+    if (refs != null) {
+      for (ServiceReference<?> ref : refs) {
+        Object service = bundleContext.getService(ref);
```

Comment:
Do we runs this on every CQL execution? I suppose this will case OSGi reference leaks since we never "unget" :thinking:

Reply:
Added